src/maxdiffusion/train_sdxl.py

```diff
@@ -0,0 +1,513 @@
+      'perf/step_time_seconds': step_time_delta.total_seconds()
+  })
+  metrics['scalar'].update({
+      'perf/per_device_tflops' : per_device_tflops
+  })
+  metrics['scalar'].update({
+      'perf/per_device_tflops_per_sec':
+          per_device_tflops /
+          step_time_delta.total_seconds()
+  })
+  metrics['scalar'].update({'learning/current_learning_rate': lr })
+
+_buffered_step = None
+_buffered_metrics = None
+def write_metrics(writer, local_metrics_file, running_gcs_metrics, metrics, step, config):
+  """Entry point for all metrics writing in Train's Main.
+     TODO: would be better as a Class in the future (that initialized all state!)
+
+     To avoid introducing an unnecessary dependency, we "double buffer" -- we hold
+     onto the last metrics and step and only publish when we receive a new metrics and step.
+     The logic is that this ensures that Jax is able to queues train_steps and we
+     don't block when turning "lazy" Jax arrays into real Python numbers.
+  """
+  global _buffered_step, _buffered_metrics
+
+  if _buffered_metrics is not None:
+    if _buffered_step is None:
+      raise ValueError(f"When writing metrics, {_buffered_step=} was none")
+    write_metrics_to_tensorboard(writer, _buffered_metrics, _buffered_step, config)
+
+    if config.metrics_file:
+      max_utils.write_metrics_locally(_buffered_metrics, _buffered_step, config, local_metrics_file)
+
+    if config.gcs_metrics and jax.process_index() == 0:
+      running_gcs_metrics = max_utils.write_metrics_for_gcs(_buffered_metrics, _buffered_step, config, running_gcs_metrics)
+
+  _buffered_step = step
+  _buffered_metrics = metrics
+
+def write_metrics_to_tensorboard(writer, metrics, step, config):
+  """Writes metrics to tensorboard"""
+  with jax.spmd_mode('allow_all'):
+    if jax.process_index() == 0:
+      for metric_name in metrics.get("scalar",[]):
+        writer.add_scalar(metric_name, np.array(metrics["scalar"][metric_name]), step)
+      for metric_name in metrics.get("scalars",[]):
+        writer.add_scalars(metric_name, metrics["scalars"][metric_name], step)
+
+    full_log = step % config.log_period == 0
+    if jax.process_index() == 0:
+        max_logging.log(f"completed step: {step}, seconds: {metrics['scalar']['perf/step_time_seconds']:.3f}, "
+            f"TFLOP/s/device: {metrics['scalar']['perf/per_device_tflops_per_sec']:.3f}, "
+            f"loss: {metrics['scalar']['learning/loss']:.3f}")
+
+    if full_log and jax.process_index() == 0:
+      max_logging.log(
+          f"To see full metrics 'tensorboard --logdir={config.tensorboard_dir}'"
+      )
+      writer.flush()
+
+def calculate_num_params_from_pytree(params):
+  params_sizes = jax.tree_util.tree_map(jax.numpy.size, params)
+  total_parameters = jax.tree_util.tree_reduce(lambda x, y: x + y, params_sizes)
+  return total_parameters
+
+def get_params_to_save(params):
+    return jax.device_get(jax.tree_util.tree_map(lambda x: x, params))
```

Comment:
Up until here, can be used for both SD 2 training and SDXL training, can we move them to another file/ appropriate existing files and use them instead of copy pasting?